add 2.0 upgrade guide (#608)
Azure/terraform-provider-azapi: docs/guides/2.0-upgrade-guide.md

---
layout: "azapi"
page_title: "AzAPI Provider: 2.0 Upgrade Guide"
description: |-
  This guide will cover how to upgrade from version 1.x to version 2.0 of the AzAPI Provider.

---

# Azapi Provider

## Azapi Provider Version v2.0

### Considerations

Version 2.0 of the Azapi Provider is a major release and includes breaking changes which are outlined in this document.

When upgrading to version 2.0 of the Azure Provider, we recommend upgrading to the latest version of Terraform Core ([which can be found here](https://www.terraform.io/downloads)).

This guide will continue to receive updates as we incorporate feedback and continue to fix any issues during the Beta.

### Pinning your Provider Version

We recommend pinning the version of each Provider you use in Terraform - you can do this using the `version` attribute within the `required_providers` block, either to a specific version of the Azapi Provider, like so:

```hcl
terraform {
  required_providers {
    azapi = {
      source  = "azure/azapi"
      version = "=1.15.0"
    }
  }
}

provider "azapi" {
}
```

.. or to any 1.x release:

```hcl
terraform {
  required_providers {

>     azurerm = {

Reviewer: I think this should be "azapi" instead of "azurerm"

      source  = "azure/azapi"
      version = "~> 1.x"
    }
  }
}

provider "azapi" {
}
```

More information on [how to pin the version of a Terraform Provider being used can be found on the Terraform Website](https://www.terraform.io/language/providers/requirements#requiring-providers).

---

## What's available in Version 2.0 of the Azapi Provider?

Below is an overview of the changes coming 2.0. Each topic is covered in more detail further down.

* [Dynamic Properties Support](#dynamic-properties-support)
* [Provider Functions](#provider-functions)

> * [Cusomtized Retry Configuration](#customized-retry-configuration)

Reviewer 2: ```suggestion
* [Customized Retriable Configuration](#customized-retry-configuration)
```

The dynamic properties support makes it much easier to define resources and data sources in the AzAPI Provider. This feature is available for all resources and data sources in the AzAPI Provider.

```hcl
resource "azapi_resource" "automationAccount" {
  type      = "Microsoft.Automation/automationAccounts@2023-11-01"
  parent_id = azapi_resource.resourceGroup.id
  name      = "example-automation-account"
  location  = "westeurope"

  // it's not necessary to use `jsonencode` function to encode the HCL object to JSON, just use the HCL object directly
  body = {
    properties = {
      encryption = {
        keySource = "Microsoft.Automation"
      }
      publicNetworkAccess = true
      sku = {
        name = "Basic"
      }
    }
  }
  response_export_values = ["properties"]

  // native lifecycle management
  lifecycle {
    ignore_changes = [
      body.properties.encryption.keySource
    ]
  }
}

output "o1" {
  // it's not necessary to use `jsondecode` function to decode the response
  value = azapi_resource.automationAccount.output.properties.automationHybridServiceUrl
}
```

> 

Reviewer 2: ``
```suggestion
For more information, refer to [this blogpost](https://techcommunity.microsoft.com/t5/azure-tools-blog/announcing-azapi-dynamic-properties/ba-p/4121855)
```

   type      = "Microsoft.ContainerRegistry/registries@2020-11-01-preview"
   parent_id = "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/example"
    name      = "example"
    response_export_values = {
      login_server = "properties.loginServer"
      quarantine_status = "properties.policies.quarantinePolicy.status"
    }
}

// it will output below object
// {
//   "login_server" = "example.azurecr.io"
//   "quarantine_status" = "Enabled" 
// }

output "login_server" {
  // it will output "example.azurecr.io"
  value = data.azapi_resource.example.output.login_server
}

output "quarantine_status" {
  // it will output "Enabled"
  value = data.azapi_resource.example.output.quarantine_status
}

```

## Breaking Changes

- Provider field `default_naming_prefix` and `default_naming_suffix` are removed. 

  How to fix:  
  Please remove the `default_naming_prefix` and `default_naming_suffix` fields from the provider block. You can specify the naming prefix and suffix in the resource's `name` field instead.

- Provider field `enable_hcl_output_for_data_source` is removed. The `output` field in the data source is always in HCL format.

  How to fix:  

>   Please remove the `enable_hcl_output_for_data_source` field from the provider block. And remove the `jsondecode` function when using the `output` field in the data source.

Reviewer 2: ```suggestion
  Please remove the `enable_hcl_output_for_data_source` field from the provider block. Remove any `jsondecode` functions when using the `output` field in the data source.
```

- The `azapi_resource`'s `removing_special_chars` field is removed. Please specify the `name` field and remove the special characters in the `name` field instead.

  How to fix:  

>   Please remove the `removing_special_chars` field from the `azapi_resource` block. And remove the special characters in the `name` field.

Reviewer 2: ```suggestion
  Please remove the `removing_special_chars` field from the `azapi_resource` block and any special characters in the `name` field.
```
